File: python/problems/sliding_window/longest_substring_without_repeating_characters.py
"""Longest Substring Without Repeating Characters.

Return the length of the longest consecutive part of s with no repeated
characters.

n = number of characters in s.

How to test:
Test file: tests/python/test_longest_substring_without_repeating_characters.py
Run: PYTHONDONTWRITEBYTECODE=1 python3 -m unittest -v tests/python/test_longest_substring_without_repeating_characters.py
"""


def length_of_longest_substring_brute_force(s: str) -> int:
    """Begin every possible substring and stop when it repeats a character.

    Time: O(n^2), because each starting character may scan many later characters.
    Space: O(n), for the set that tracks one substring.
    """
    # Store the longest unique substring length found so far.
    longest_length = 0

    # Try every character as the beginning of a substring.
    for start in range(len(s)):
        # Track characters in the substring that begins at start.
        characters = set()

        # Extend this substring one character at a time.
        for end in range(start, len(s)):
            # Read the character that is trying to join this substring.
            char = s[end]

            # A duplicate means this substring cannot grow further.
            if char in characters:
                break

            # Add the new unique character to this substring.
            characters.add(char)
            # end - start + 1 is this substring's length.
            longest_length = max(longest_length, end - start + 1)

    # Return the largest unique substring length.
    return longest_length


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """Use a sliding window and shrink it only when a duplicate appears.

        Time: O(n), because each character enters once and leaves at most once.
        Space: O(n), for the set holding current window characters.
        """
        # left is the first index of the current unique window.
        left = 0
        # Store the longest valid window length found so far.
        longest_length = 0
        # Store characters currently inside the window.
        characters = set()

        # right expands the window by reading every character once.
        for right in range(len(s)):
            # Read the character that is trying to enter the window.
            char = s[right]

            # Shrink from the left until char is no longer duplicated.
            while char in characters:
                # Remove the character that is leaving the window.
                characters.remove(s[left])
                # Move the start of the window rightward.
                left += 1

            # Add char after the window no longer contains a duplicate.
            characters.add(char)
            # right - left + 1 is the current valid window length.
            longest_length = max(longest_length, right - left + 1)

        # Return the largest unique window length.
        return longest_length

File: javascript/problems/sliding_window/longestSubstringWithoutRepeatingCharacters.js
/**
 * Longest Substring Without Repeating Characters
 *
 * Return the length of the longest consecutive part of s with no repeated
 * characters.
 *
 * n = number of characters in s.
 *
 * How to test:
 * Test file: tests/javascript/longestSubstringWithoutRepeatingCharacters.test.js
 * Run: node --test tests/javascript/longestSubstringWithoutRepeatingCharacters.test.js
 */

/**
 * Brute force: begin every possible substring and stop when it repeats.
 *
 * Time: O(n^2), because each starting character may scan many later characters.
 * Space: O(n), for the Set that tracks one substring.
 */
function lengthOfLongestSubstringBruteForce(s) {
  // Store the longest unique substring length found so far.
  let longestLength = 0;

  // Try every character as the starting point of a substring.
  for (let start = 0; start < s.length; start += 1) {
    // Track characters in the substring beginning at start.
    const characters = new Set();

    // Extend this substring one character at a time.
    for (let end = start; end < s.length; end += 1) {
      // Read the new character at the end of this substring.
      const char = s[end];

      // A repeated character means this substring cannot grow further.
      if (characters.has(char)) {
        break;
      }

      // Add the new unique character to this substring.
      characters.add(char);
      // Update the longest length using the two indexes.
      longestLength = Math.max(longestLength, end - start + 1);
    }
  }

  // Return the largest unique substring length.
  return longestLength;
}

/**
 * Optimized sliding window: move left only when a duplicate enters the window.
 *
 * Time: O(n), because each character is added once and removed at most once.
 * Space: O(n), for the Set holding the current window characters.
 */
function lengthOfLongestSubstring(s) {
  // left marks the beginning of the current unique window.
  let left = 0;
  // Store the best window length found so far.
  let longestLength = 0;
  // Store characters currently inside the window from left through right.
  const characters = new Set();

  // right expands the window by visiting every character once.
  for (let right = 0; right < s.length; right += 1) {
    // Read the character that is trying to enter the window.
    const char = s[right];

    // Keep shrinking from the left until char is no longer a duplicate.
    while (characters.has(char)) {
      // Remove the character that is leaving the window.
      characters.delete(s[left]);
      // Move the beginning of the window rightward.
      left += 1;
    }

    // Add char now that the window has no duplicate char.
    characters.add(char);
    // right - left + 1 is the current window length.
    longestLength = Math.max(longestLength, right - left + 1);
  }

  // Return the largest unique window length.
  return longestLength;
}

module.exports = {
  lengthOfLongestSubstring,
  lengthOfLongestSubstringBruteForce,
};
